roachtest: sqlsmith/setup=empty/setting=no-mutations failed [error when decoding tuples with arrays]

roachtest.sqlsmith/setup=empty/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3804493&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3804493&tab=artifacts#/sqlsmith/setup=empty/setting=no-mutations) on master @ [40f11fead0a0453969634f8ddb0502c1f78b2806](https://github.com/cockroachdb/cockroach/commits/40f11fead0a0453969634f8ddb0502c1f78b2806):


```
SELECT
  t2.c4
FROM
  (VALUES (NULL, NULL), (NULL, 0::INT8)) AS t1 (c1, c2),
  (
    VALUES
      (NULL, (1:::DECIMAL, ARRAY[]:::RECORD[])),
      (NULL, (2:::DECIMAL, ARRAY[(1::INT8, 2::INT8)]))
  )
    AS t2 (c3, c4)
ORDER BY
  t1.c1 ASC;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.sqlsmith/setup=empty/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3824348&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3824348&tab=artifacts#/sqlsmith/setup=empty/setting=no-mutations) on master @ [a80cfbee826f70988381a0a85c0fe7aba0115484](https://github.com/cockroachdb/cockroach/commits/a80cfbee826f70988381a0a85c0fe7aba0115484):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
	sqlsmith.go:224,sqlsmith.go:257,test_runner.go:779: error: pq: internal error: unexpected error from the vectorized engine: error decoding 36 bytes: couldn't decode type anyelement
		stmt:
		SELECT
			tab_321530.col_555878 AS col_555882
		FROM
			(
				VALUES
					(NULL),
					(NULL),
					((ARRAY[231049373:::OID], '1':::STRING, (-24584):::INT8, ARRAY[]:::RECORD[])),
					(
						(ARRAY[2331024460:::OID,2534714298:::OID,1089071710:::OID,2169848099:::OID,389203509:::OID], 'o':::STRING, 23701:::INT8, ARRAY[]:::RECORD[])
					),
					((ARRAY[478187369:::OID,3847926774:::OID,0:::OID], e'\r':::STRING, (-23782):::INT8, ARRAY[(),(),(),()])),
					(NULL)
			)
				AS tab_321530 (col_555878)
		WHERE
			(true OR ('\x8d593fad19e6':::BYTES::BYTES NOT IN (WITH with_101239 (col_555879) AS (SELECT * FROM (VALUES (e't\x13U#Sl/c':::STRING), ('K':::STRING), (('63f4:eb15:46bd:9ddd:73b1:c75b:6728:7e81/95':::INET::INET || '':::STRING::STRING)::STRING), (e'\r\x19Ms\x1b$':::STRING), (e'>Qf.e\x061\x01':::STRING)) AS tab_321531 (col_555879)), with_101240 (col_555880) AS (SELECT * FROM (VALUES (NULL), (e'\x02\x1bYL6<':::STRING), (e'\U00002603':::STRING), (e'\x1e5v\x19\x01U\x0e\x15':::STRING), (e'I\'O88X5':::STRING)) AS tab_321532 (col_555880)) SELECT '\x3da8':::BYTES::BYTES AS col_555881 LIMIT 76:::INT8)))
		ORDER BY
			tab_321530.col_555878;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -254,6 +254,11 @@ func (v *distSQLExprCheckVisitor) VisitPre(expr tree.Expr) (recurse bool, newExp
 			v.err = newQueryNotSupportedErrorf("array %s cannot be executed with distsql", t)
 			return false, expr
 		}
+	case *tree.DTuple:
+		if t.ResolvedType() == types.AnyTuple {
+			v.err = newQueryNotSupportedErrorf("tuple %s cannot be executed with distsql", t)
+			return false, expr
+		}
 	}
 	return true, expr
 }

--- a/pkg/sql/logictest/testdata/logic_test/tuple
+++ b/pkg/sql/logictest/testdata/logic_test/tuple
@@ -1084,3 +1084,41 @@ ORDER BY
 ----
 {}
 {"(1,cat)"}
+
+query T
+SELECT
+  t2.c4
+FROM
+  (
+    VALUES
+      (NULL, (1:::DECIMAL, ARRAY[]:::record[])),
+      (NULL, (2:::DECIMAL, ARRAY[(1::INT8, 2::INT8)]))
+  )
+    AS t2 (c3, c4)
+----
+(1,{})
+(2,"{""(1,2)""}")
+
+statement error VALUES types tuple\{decimal, tuple\{int, int\}\[\]\} and tuple\{decimal, tuple\{int, int, int\}\[\]\} cannot be matched
+SELECT
+  t2.c4
+FROM
+  (
+    VALUES
+      (NULL, (1:::DECIMAL, ARRAY[]:::record[])),
+      (NULL, (2:::DECIMAL, ARRAY[(1::INT8, 2::INT8, 3::INT8)])),
+      (NULL, (2:::DECIMAL, ARRAY[(1::INT8, 2::INT8)]))
+  )
+    AS t2 (c3, c4)
+
+statement error VALUES types tuple\{decimal, tuple\{int, int, int\}\[\]\} and tuple\{decimal, tuple\{int, int\}\[\]\} cannot be matched
+SELECT
+  t2.c4
+FROM
+  (
+    VALUES
+      (NULL, (1:::DECIMAL, ARRAY[]:::record[])),
+      (NULL, (2:::DECIMAL, ARRAY[(1::INT8, 2::INT8)])),
+      (NULL, (2:::DECIMAL, ARRAY[(1::INT8, 2::INT8, 3::INT8)]))
+  )
+    AS t2 (c3, c4)

--- a/pkg/sql/opt/optbuilder/values.go
+++ b/pkg/sql/opt/optbuilder/values.go
@@ -78,14 +78,10 @@ func (b *Builder) buildValuesClause(
 			if typ := texpr.ResolvedType(); typ.Family() != types.UnknownFamily {
 				if colTypes[colIdx].Family() == types.UnknownFamily {
 					colTypes[colIdx] = typ
-				} else if colTypes[colIdx].Family() == types.ArrayFamily &&
-					colTypes[colIdx].ArrayContents() == types.AnyTuple &&
-					typ.Family() == types.ArrayFamily &&
-					typ.ArrayContents().Family() == types.TupleFamily &&
-					typ.ArrayContents() != types.AnyTuple {
-					// This more complicated condition handles the case when an earlier
-					// expression in the VALUES clause is an array of AnyTuple, but a
-					// later expression is an array of a more specific tuple type.
+				} else if rightHasMoreSpecificTuple(colTypes[colIdx], typ) {
+					// This condition handles the case when an earlier expression in the
+					// VALUES clause is an array of AnyTuple, but a later expression is an
+					// array of a more specific tuple type.
 					colTypes[colIdx] = typ
 				} else if !typ.Equivalent(colTypes[colIdx]) {
 					panic(pgerror.Newf(pgcode.DatatypeMismatch,
@@ -140,4 +136,39 @@ func reportValuesLenError(expected, actual int) {
 		pgcode.Syntax,
 		"VALUES lists must all be the same length, expected %d columns, found %d",
 		expected, actual))
+}
+
+// rightHasMoreSpecificTuple returns true if the left and right types are
+// equivalent, but the right type is constrained by a more specific nested
+// tuple than the left type.
+func rightHasMoreSpecificTuple(left, right *types.T) bool {
+	if left.Family() != right.Family() {
+		return false
+	}
+	if left.Family() == types.ArrayFamily && right.Family() == types.ArrayFamily {
+		return rightHasMoreSpecificTuple(left.ArrayContents(), right.ArrayContents())
+	}
+	if left.Family() == types.TupleFamily && right.Family() == types.TupleFamily {
+		if right == types.AnyTuple {
+			return false
+		} else if left == types.AnyTuple {
+			return true
+		} else if len(left.TupleContents()) != len(right.TupleContents()) {
+			return false
+		}
+		ret := true
+		for i, leftElem := range left.TupleContents() {
+			rightElem := right.TupleContents()[i]
+			// Only recurse if we are dealing with a container type.
+			if leftElem.Family() == types.ArrayFamily || leftElem.Family() == types.TupleFamily {
+				ret = ret && rightHasMoreSpecificTuple(leftElem, rightElem)
+			} else {
+				ret = ret && leftElem.Equivalent(rightElem)
+			}
+		}
+		return ret
+	}
+	// For non-tuples and non-arrays, we don't want the right type to get
+	// preference over the left type.
+	return false
 }

--- a/pkg/sql/sem/tree/walk.go
+++ b/pkg/sql/sem/tree/walk.go
@@ -780,10 +780,24 @@ func (expr *DTimestamp) Walk(_ Visitor) Expr { return expr }
 func (expr *DTimestampTZ) Walk(_ Visitor) Expr { return expr }
 
 // Walk implements the Expr interface.
-func (expr *DTuple) Walk(_ Visitor) Expr { return expr }
+func (expr *DTuple) Walk(v Visitor) Expr {
+	for _, d := range expr.D {
+		// Datums never get changed by visitors, so we don't need to
+		// look for changed elements.
+		d.Walk(v)
+	}
+	return expr
+}
 
 // Walk implements the Expr interface.
-func (expr *DArray) Walk(_ Visitor) Expr { return expr }
+func (expr *DArray) Walk(v Visitor) Expr {
+	for _, d := range expr.Array {
+		// Datums never get changed by visitors, so we don't need to
+		// look for changed elements.
+		d.Walk(v)
+	}
+	return expr
+}
 
 // Walk implements the Expr interface.
 func (expr *DOid) Walk(_ Visitor) Expr { return expr }

